shim/src/main.rs

```diff
@@ -0,0 +1,144 @@
+mod package_manager;
+mod paths;
+
+use crate::package_manager::PackageManager;
+use crate::paths::AncestorSearch;
+use anyhow::Result;
+use clap::{Parser, Subcommand};
+use std::path::Path;
+use std::{
+    env,
+    ffi::CString,
+    os::raw::{c_char, c_int},
+    process,
+};
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None, ignore_errors = true, disable_help_flag = true)]
+struct Args {
+    /// Current working directory
+    #[clap(long, value_parser)]
+    cwd: Option<String>,
+    #[clap(subcommand)]
+    commands: Option<Commands>,
+    task: Option<String>,
+}
+
+/// Defines the subcommands for CLI. NOTE: If we change the commands in Go,
+/// we must change these as well to avoid accidentally passing the --single-package
+/// flag into non-build commands.
+#[derive(Subcommand, Debug)]
+enum Commands {
+    Bin,
+    Completion,
+    Daemon,
+    Help,
+    Link,
+    Login,
+    Logout,
+    Prune,
+    Unlink,
+    Run { tasks: Vec<String> },
+}
+
+extern "C" {
+    pub fn nativeRunWithArgs(argc: c_int, argv: *mut *mut c_char) -> c_int;
+}
+
+/// Runs the turbo in the current binary
+///
+/// # Arguments
+///
+/// * `args`: Arguments for turbo
+///
+/// returns: Result<i32, Error>
+///
+fn run_current_turbo(args: Vec<String>) -> Result<i32> {
+    let mut args = args
+        .into_iter()
+        .map(|s| {
+            let c_string = CString::new(s)?;
+            Ok(c_string.into_raw())
+        })
+        .collect::<Result<Vec<*mut c_char>>>()?;
+    args.shrink_to_fit();
+    let argc: c_int = args.len() as c_int;
+    let argv = args.as_mut_ptr();
+    let exit_code = unsafe { nativeRunWithArgs(argc, argv) };
+    Ok(exit_code)
+}
+
+/// Checks if we are in single package mode by first seeing if there is a turbo.json
+/// in the ancestor path, and then checking for workspaces.
+///
+/// # Arguments
+///
+/// * `current_dir`: Current working directory
+///
+/// returns: Result<bool, Error>
+///
+fn is_single_package_mode(current_dir: &Path) -> Result<bool> {
+    let has_turbo_json = AncestorSearch::new(current_dir.to_path_buf(), "turbo.json")?
+        .next()
+        .is_some();
+
+    if has_turbo_json {
+        return Ok(false);
+    }
```

Comment:
Curious... Single Package Mode doesn't have support for a `turbo.json` at all? Where would one declare `globalDependencies`, `globalEnv`?

Reply:
Single-package mode should support `turbo.json`, but it isn't required.

I think this logic here is probably wrong. Apologies if I missed that on an earlier review.

Reply:
Alright, interesting. So we determine single package mode based on the workspaces, and not on `turbo.json`, correct? Like if we detect workspaces in either `package.json` or `pnpm-workspace.yml`, then we are in multi-package, otherwise single-package.

Reply:
Yeah, I think this is the "ancestor path contains single `turbo.json`" case, where we need to check whether or not workspaces are defined.

Reply:
I'm going to cherry-pick the local turbo logic onto this PR because that is effectively a rewrite of this logic. Is multiple turbo.json a thing already?